File: win/directx/sdk-9.0/Samples/C++/Direct3D/BumpMapping/DotProduct3/dotproduct3.cpp
//-----------------------------------------------------------------------------
// File: DotProduct3.cpp
//
// Desc: D3D sample showing how to do bumpmapping using the DotProduct3 
//       texture operation.
// 
// Copyright (c) Microsoft Corporation. All rights reserved.
//-----------------------------------------------------------------------------
#define STRICT
#include <Windows.h>
#include <commctrl.h>
#include <tchar.h>
#include <math.h>
#include <stdio.h>
#include <D3DX9.h>
#include "DXUtil.h"
#include "D3DEnumeration.h"
#include "D3DSettings.h"
#include "D3DApp.h"
#include "D3DFont.h"
#include "D3DUtil.h"
#include "resource.h"




//-----------------------------------------------------------------------------
// Name: 
// Desc: 
//-----------------------------------------------------------------------------
struct CUSTOMVERTEX
{
    D3DXVECTOR3 v;
    DWORD       diffuse;
    DWORD       specular;
    FLOAT       tu, tv;

    static const DWORD FVF;
};
const DWORD CUSTOMVERTEX::FVF = D3DFVF_XYZ | D3DFVF_DIFFUSE | D3DFVF_SPECULAR | D3DFVF_TEX1;



//-----------------------------------------------------------------------------
// Name: class CMyD3DApplication
// Desc: Application class. The base class (CD3DApplication) provides the 
//       generic functionality needed in all Direct3D samples. CMyD3DApplication 
//       adds functionality specific to this sample program.
//-----------------------------------------------------------------------------
class CMyD3DApplication : public CD3DApplication
{
    CD3DFont*          m_pFont;
    CUSTOMVERTEX       m_QuadVertices[4];
    LPDIRECT3DTEXTURE9 m_pCustomNormalMap;
    LPDIRECT3DTEXTURE9 m_pFileBasedNormalMap;
    D3DXVECTOR3        m_vLight;

    BOOL               m_bUseFileBasedTexture;
    BOOL               m_bShowNormalMap;

    HRESULT CreateFileBasedNormalMap();
    HRESULT CreateCustomNormalMap();
    HRESULT ConfirmDevice( D3DCAPS9*, DWORD, D3DFORMAT, D3DFORMAT );
    LRESULT MsgProc( HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lParam );

protected:
    HRESULT OneTimeSceneInit();
    HRESULT InitDeviceObjects();
    HRESULT RestoreDeviceObjects();
    HRESULT InvalidateDeviceObjects();
    HRESULT DeleteDeviceObjects();
    HRESULT Render();
    HRESULT FrameMove();
    HRESULT FinalCleanup();

public:
    CMyD3DApplication();
};




//-----------------------------------------------------------------------------
// Name: WinMain()
// Desc: Entry point to the program. Initializes everything, and goes into a
//       message-processing loop. Idle time is used to render the scene.
//-----------------------------------------------------------------------------
INT WINAPI WinMain( HINSTANCE hInst, HINSTANCE, LPSTR, INT )
{
    CMyD3DApplication d3dApp;

    InitCommonControls();
    if( FAILED( d3dApp.Create( hInst ) ) )
        return 0;

    return d3dApp.Run();
}




//-----------------------------------------------------------------------------
// Name: VectortoRGBA()
// Desc: Turns a normalized vector into RGBA form. Used to encode vectors into
//       a height map. 
//-----------------------------------------------------------------------------
DWORD VectortoRGBA( D3DXVECTOR3* v, FLOAT fHeight )
{
    DWORD r = (DWORD)( 127.0f * v->x + 128.0f );
    DWORD g = (DWORD)( 127.0f * v->y + 128.0f );
    DWORD b = (DWORD)( 127.0f * v->z + 128.0f );
    DWORD a = (DWORD)( 255.0f * fHeight );
    
    return( (a<<24L) + (r<<16L) + (g<<8L) + (b<<0L) );
}




//-----------------------------------------------------------------------------
// Name: InitVertex()
// Desc: Initializes a vertex
//-----------------------------------------------------------------------------
VOID InitVertex( CUSTOMVERTEX* vtx, FLOAT x, FLOAT y, FLOAT z, FLOAT tu, FLOAT tv )
{
    D3DXVECTOR3 v(1,1,1);
    D3DXVec3Normalize( &v, &v );
    vtx[0].v        = D3DXVECTOR3( x, y, z );
    vtx[0].diffuse  = VectortoRGBA( &v, 1.0f );
    vtx[0].specular = 0x40400000;
    vtx[0].tu       = tu;
    vtx[0].tv       = tv;
}
    



//-----------------------------------------------------------------------------
// Name: CMyD3DApplication()
// Desc: Application constructor. Sets attributes for the app.
//-----------------------------------------------------------------------------
CMyD3DApplication::CMyD3DApplication()
{
    m_strWindowTitle  = _T("DotProduct3: BumpMapping Technique");
    m_d3dEnumeration.AppUsesDepthBuffer = FALSE;

    m_pFont                = new CD3DFont( _T("Arial"), 12, D3DFONT_BOLD );

    m_bUseFileBasedTexture = FALSE;
    m_bShowNormalMap       = FALSE;

    m_pCustomNormalMap     = NULL;
    m_pFileBasedNormalMap  = NULL;
}




//-----------------------------------------------------------------------------
// Name: OneTimeSceneInit()
// Desc: Called during initial app startup, this function performs all the
//       permanent initialization.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::OneTimeSceneInit()
{
    InitVertex( &m_QuadVertices[0],-1.0f,-1.0f,-1.0f, 0.0f, 0.0f );
    InitVertex( &m_QuadVertices[1], 1.0f,-1.0f,-1.0f, 1.0f, 0.0f );
    InitVertex( &m_QuadVertices[2],-1.0f, 1.0f,-1.0f, 0.0f, 1.0f );
    InitVertex( &m_QuadVertices[3], 1.0f, 1.0f,-1.0f, 1.0f, 1.0f );

    m_vLight = D3DXVECTOR3( 0.0f, 0.0f, 1.0f );

    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: FrameMove()
// Desc: Called once per frame, the call is the entry point for animating
//       the scene.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::FrameMove()
{
    // Compute the light vector from the cursor position
    if( GetFocus() )
    {
        POINT pt;
        GetCursorPos( &pt );
        ScreenToClient( m_hWnd, &pt );

        m_vLight.x = -( ( ( 2.0f * pt.x ) / m_d3dsdBackBuffer.Width ) - 1 );
        m_vLight.y = -( ( ( 2.0f * pt.y ) / m_d3dsdBackBuffer.Height ) - 1 );
        m_vLight.z = 0.0f;

        if( D3DXVec3Length( &m_vLight ) > 1.0f )
            D3DXVec3Normalize( &m_vLight, &m_vLight );
        else
            m_vLight.z = sqrtf( 1.0f - m_vLight.x*m_vLight.x
                                     - m_vLight.y*m_vLight.y );
    }

    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: Render()
// Desc: Called once per frame, the call is the entry point for 3d
//       rendering. This function sets up render states, clears the
//       viewport, and renders the scene.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::Render()
{
    // Clear the render target
    m_pd3dDevice->Clear( 0L, NULL, D3DCLEAR_TARGET, 0x00000000f, 1.0f, 0L );

    // Begin the scene
    if( SUCCEEDED( m_pd3dDevice->BeginScene() ) )
    {
        // Store the light vector, so it can be referenced in D3DTA_TFACTOR
        DWORD dwFactor = VectortoRGBA( &m_vLight, 0.0f );
        m_pd3dDevice->SetRenderState( D3DRS_TEXTUREFACTOR, dwFactor );

        // Modulate the texture (the normal map) with the light vector (stored
        // above in the texture factor)
        m_pd3dDevice->SetTextureStageState( 0, D3DTSS_COLORARG1, D3DTA_TEXTURE );
        m_pd3dDevice->SetTextureStageState( 0, D3DTSS_COLOROP,   D3DTOP_DOTPRODUCT3 );
        m_pd3dDevice->SetTextureStageState( 0, D3DTSS_COLORARG2, D3DTA_TFACTOR );

        // If user wants to see the normal map, override the above renderstates and
        // simply show the texture
        if( TRUE == m_bShowNormalMap )
            m_pd3dDevice->SetTextureStageState( 0, D3DTSS_COLOROP, D3DTOP_SELECTARG1 );

        // Select which normal map to use
        if( m_bUseFileBasedTexture )
            m_pd3dDevice->SetTexture( 0, m_pFileBasedNormalMap );
        else
            m_pd3dDevice->SetTexture( 0, m_pCustomNormalMap );

        // Draw the bumpmapped quad
        m_pd3dDevice->SetFVF( CUSTOMVERTEX::FVF );
        m_pd3dDevice->DrawPrimitiveUP( D3DPT_TRIANGLESTRIP, 2, m_QuadVertices, 
                                       sizeof(CUSTOMVERTEX) );

        // Output statistics
        m_pFont->DrawText( 2,  0, D3DCOLOR_ARGB(255,255,255,0), m_strFrameStats );
        m_pFont->DrawText( 2, 20, D3DCOLOR_ARGB(255,255,255,0), m_strDeviceStats );

        // End the scene.
        m_pd3dDevice->EndScene();
    }

    return S_OK;
}





//-----------------------------------------------------------------------------
// Name: 
// Desc: 
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::CreateFileBasedNormalMap()
{
    HRESULT hr = S_OK;
    LPDIRECT3DTEXTURE9 pFileBasedNormalMapSource = NULL;

    // Load the texture from a file
    if( FAILED( hr = D3DUtil_CreateTexture( m_pd3dDevice, _T("EarthBump.bmp"), 
                                            &pFileBasedNormalMapSource, 
                                            D3DFMT_A8R8G8B8 ) ) )
    {
        return D3DAPPERR_MEDIANOTFOUND;
    }

    D3DSURFACE_DESC desc;
    pFileBasedNormalMapSource->GetLevelDesc( 0, &desc );    

    if( FAILED( hr = D3DXCreateTexture( m_pd3dDevice, desc.Width, desc.Height, 
        pFileBasedNormalMapSource->GetLevelCount(), 0, 
        D3DFMT_A8R8G8B8, D3DPOOL_MANAGED, &m_pFileBasedNormalMap ) ) )
    {
        goto End;
    }

    if( FAILED( hr = D3DXComputeNormalMap( m_pFileBasedNormalMap, pFileBasedNormalMapSource,
        NULL, 0, D3DX_CHANNEL_RED, 1.0f ) ) )
    {
        goto End;
    }

End:
    SAFE_RELEASE( pFileBasedNormalMapSource );
    return hr;
}




//-----------------------------------------------------------------------------
// Name: 
// Desc: 
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::CreateCustomNormalMap()
{
    DWORD   dwWidth  = 512;
    DWORD   dwHeight = 512;
    HRESULT hr;

    // Create a 32-bit texture for the custom normal map
    hr = m_pd3dDevice->CreateTexture( dwWidth, dwHeight, 1, 
                                      0 /* Usage */, 
                                      D3DFMT_A8R8G8B8, D3DPOOL_MANAGED, 
                                      &m_pCustomNormalMap, NULL );
    if( FAILED(hr) )
        return hr;

    // Lock the texture to fill it with our custom image
    D3DLOCKED_RECT d3dlr;
    if( FAILED( m_pCustomNormalMap->LockRect( 0, &d3dlr, 0, 0 ) ) )
        return E_FAIL;
    DWORD* pPixel = (DWORD*)d3dlr.pBits;

    // Fill each pixel
    for( DWORD j=0; j<dwHeight; j++  )
    {
        for( DWORD i=0; i<dwWidth; i++ )
        {
            FLOAT xp = ( (5.0f*i) / (dwWidth-1)  );
            FLOAT yp = ( (5.0f*j) / (dwHeight-1) );
            FLOAT x  = 2*(xp-floorf(xp))-1;
            FLOAT y  = 2*(yp-floorf(yp))-1;
            FLOAT z  = sqrtf( 1.0f - x*x - y*y );

            // Make image of raised circle. Outside of circle is gray
            if( (x*x + y*y) <= 1.0f )
            {
                D3DXVECTOR3 vVector( x, y, z );
                *pPixel++ = VectortoRGBA( &vVector, 1.0f );
            }
            else
                *pPixel++ = 0x80808080;
        }
    }

    // Unlock the map and return successful
    m_pCustomNormalMap->UnlockRect(0);

    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: InitDeviceObjects()
// Desc: Initialize scene objects.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::InitDeviceObjects()
{
    HRESULT hr;

    m_pFont->InitDeviceObjects( m_pd3dDevice );

    // Create the normal maps
    if( FAILED( hr = CreateFileBasedNormalMap() ) )
        return hr;
    if( FAILED( hr = CreateCustomNormalMap() ) )
        return hr;

    // Set menu states
    CheckMenuItem( GetMenu(m_hWnd), IDM_USEFILEBASEDTEXTURE,
                   m_bUseFileBasedTexture ? MF_CHECKED : MF_UNCHECKED );
    CheckMenuItem( GetMenu(m_hWnd), IDM_USECUSTOMTEXTURE,
                   m_bUseFileBasedTexture ? MF_UNCHECKED : MF_CHECKED );
    CheckMenuItem( GetMenu(m_hWnd), IDM_SHOWNORMALMAP,    
                   m_bShowNormalMap ? MF_CHECKED : MF_UNCHECKED );

    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: RestoreDeviceObjects()
// Desc: Initialize scene objects.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::RestoreDeviceObjects()
{
    m_pFont->RestoreDeviceObjects();

    // Set the transform matrices
    D3DXVECTOR3 vEyePt    = D3DXVECTOR3( 0.0f, 0.0f, 2.0f );
    D3DXVECTOR3 vLookatPt = D3DXVECTOR3( 0.0f, 0.0f, 0.0f );
    D3DXVECTOR3 vUpVec    = D3DXVECTOR3( 0.0f, 1.0f, 0.0f );
    D3DXMATRIXA16  matWorld, matView, matProj;
    
    D3DXMatrixIdentity( &matWorld );
    D3DXMatrixLookAtLH( &matView, &vEyePt, &vLookatPt, &vUpVec );
    FLOAT fAspect = m_d3dsdBackBuffer.Width / (FLOAT)m_d3dsdBackBuffer.Height;
    D3DXMatrixPerspectiveFovLH( &matProj, D3DX_PI/4, fAspect, 1.0f, 500.0f );
    
    m_pd3dDevice->SetTransform( D3DTS_WORLD,      &matWorld );
    m_pd3dDevice->SetTransform( D3DTS_VIEW,       &matView );
    m_pd3dDevice->SetTransform( D3DTS_PROJECTION, &matProj );
    
    // Set misc render states
    m_pd3dDevice->SetSamplerState( 0, D3DSAMP_MINFILTER, D3DTEXF_LINEAR );
    m_pd3dDevice->SetSamplerState( 0, D3DSAMP_MAGFILTER, D3DTEXF_LINEAR );
    m_pd3dDevice->SetRenderState( D3DRS_LIGHTING, FALSE );

    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: InvalidateDeviceObjects()
// Desc: Called when the app is exiting, or the device is being changed,
//       this function deletes any device dependent objects.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::InvalidateDeviceObjects()
{
    m_pFont->InvalidateDeviceObjects();
    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: DeleteDeviceObjects()
// Desc: Called when the app is exiting, or the device is being changed,
//       this function deletes any device dependent objects.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::DeleteDeviceObjects()
{
    m_pFont->DeleteDeviceObjects();
    SAFE_RELEASE( m_pFileBasedNormalMap );
    SAFE_RELEASE( m_pCustomNormalMap );

    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: FinalCleanup()
// Desc: Called before the app exits, this function gives the app the chance
//       to cleanup after itself.
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::FinalCleanup()
{
    SAFE_DELETE( m_pFont );
    return S_OK;
}




//-----------------------------------------------------------------------------
// Name: ConfirmDevice()
// Desc: Called during device initialization, this code checks the device
//       for some minimum set of capabilities
//-----------------------------------------------------------------------------
HRESULT CMyD3DApplication::ConfirmDevice( D3DCAPS9* pCaps, DWORD dwBehavior, 
                                          D3DFORMAT adapterFormat, D3DFORMAT backBufferFormat )
{
    if( pCaps->TextureOpCaps & D3DTEXOPCAPS_DOTPRODUCT3 )
        return S_OK;
    
    return E_FAIL;
}




//-----------------------------------------------------------------------------
// Name: MsgProc()
// Desc: Message proc function to handle key and menu input
//-----------------------------------------------------------------------------
LRESULT CMyD3DApplication::MsgProc( HWND hWnd, UINT uMsg, WPARAM wParam,
                                    LPARAM lParam )
{
    if( uMsg == WM_COMMAND )
    {
        switch( LOWORD(wParam) )
        {
            case IDM_USEFILEBASEDTEXTURE:
                m_bUseFileBasedTexture = TRUE;
                CheckMenuItem( GetMenu(hWnd), IDM_USEFILEBASEDTEXTURE, MF_CHECKED );
                CheckMenuItem( GetMenu(hWnd), IDM_USECUSTOMTEXTURE,    MF_UNCHECKED );
                break;
            
            case IDM_USECUSTOMTEXTURE:
                m_bUseFileBasedTexture = FALSE;
                CheckMenuItem( GetMenu(hWnd), IDM_USEFILEBASEDTEXTURE, MF_UNCHECKED );
                CheckMenuItem( GetMenu(hWnd), IDM_USECUSTOMTEXTURE,    MF_CHECKED );
                break;
            
            case IDM_SHOWNORMALMAP:
                m_bShowNormalMap = !m_bShowNormalMap;
                CheckMenuItem( GetMenu(hWnd), IDM_SHOWNORMALMAP,    
                               m_bShowNormalMap ? MF_CHECKED : MF_UNCHECKED );
                break;
        }
    }

    return CD3DApplication::MsgProc( hWnd, uMsg, wParam, lParam );
}

File: win/directx/sdk-9.0/Samples/C#/Direct3D/BumpMapping/DotProduct3/DotProduct3.cs
//-----------------------------------------------------------------------------
// File: DotProduct3.cs
//
// Desc: D3D sample showing how to do bumpmapping using the DotProduct3 
//       texture operation.
// 
// Copyright (c) Microsoft Corporation. All rights reserved.
//-----------------------------------------------------------------------------
using System;
using System.Windows.Forms;
using Microsoft.DirectX;
using Microsoft.DirectX.Direct3D;




namespace DotProductSample
{
	/// <summary>
	/// Application class. The base class (GraphicsSample) provides the generic 
	/// functionality needed in all Direct3D samples. MyGraphicsSample adds 
	/// functionality specific to this sample program.
	/// </summary>
	public class MyGraphicsSample : GraphicsSample
	{
		/// <summary>
		/// Custom Bump vertex
		/// </summary>
		public struct DotVertex
		{
			public Vector3 p;
			public int diffuse;
			public int specular;
			public float tu1, tv1;

			public static readonly VertexFormats Format = VertexFormats.Position | VertexFormats.Diffuse | VertexFormats.Specular | VertexFormats.Texture1;
		};



		private GraphicsFont drawingFont = null;                 // Font for drawing text
		// Scene
		private DotVertex[] quadVerts = new DotVertex[4]; // Quad vertex buffer data
		private Texture customNormalMap = null;
		private Texture fileBasedNormalMap = null;
		private Vector3 lightVector;

		bool isUsingFileBasedTexture = false;
		bool isShowingNormalMap = false;


		// Menu items for extra actions the sample can perform
		private MenuItem mnuOptions = null;
		private MenuItem mnuFileTexture = null;
		private MenuItem mnuCustomTexture = null;
		private MenuItem mnuBreak = null;
		private MenuItem mnuShowNormal = null;


		

		/// <summary>
		/// Application constructor. Sets attributes for the app.
		/// </summary>
		public MyGraphicsSample()
		{
			// Set the window text
			this.Text = "DotProduct3: BumpMapping Technique";
            try
            {
                // Load the icon from our resources
                System.Resources.ResourceManager resources = new System.Resources.ResourceManager(this.GetType());
                this.Icon = ((System.Drawing.Icon)(resources.GetObject("$this.Icon")));
            }
            catch
            {
                // It's no big deal if we can't load our icons, but try to load the embedded one
                try { this.Icon = new System.Drawing.Icon(this.GetType(), "directx.ico"); } 
                catch {}
            }

			drawingFont = new GraphicsFont("Arial", System.Drawing.FontStyle.Bold);
			enumerationSettings.AppUsesDepthBuffer = false;

			// Add the options menu to the main menu
			mnuOptions = new MenuItem("&Options");
			this.mnuMain.MenuItems.Add(mnuOptions);

			// Now create the rest of the menu items
			mnuFileTexture = new MenuItem("Use file-based texture");
			mnuFileTexture.Shortcut = Shortcut.CtrlF;
			mnuFileTexture.ShowShortcut = true;
			mnuFileTexture.Click += new System.EventHandler(this.FileTextureClicked);

			mnuCustomTexture = new MenuItem("Use custom texture");
			mnuCustomTexture.Shortcut = Shortcut.CtrlC;
			mnuCustomTexture.ShowShortcut = true;
			mnuCustomTexture.Click += new System.EventHandler(this.CustomTextureClicked);

			mnuBreak = new MenuItem("-");
			mnuShowNormal = new MenuItem("Show normal map");
			mnuShowNormal.Shortcut = Shortcut.CtrlN;
			mnuShowNormal.ShowShortcut = true;
			mnuShowNormal.Click += new System.EventHandler(this.ShowNormalClicked);

			// Now add the new items to the option menu
			mnuOptions.MenuItems.AddRange(new MenuItem[] { mnuFileTexture, mnuCustomTexture, mnuBreak, mnuShowNormal });
		}




		/// <summary>
		/// Turns a normalized vector into RGBA form. Used to encode vectors into a height map. 
		/// </summary>
		/// <param name="v">Normalized Vector</param>
		/// <param name="height">Height</param>
		/// <returns>RGBA Form of the vector</returns>
		private int VectorToRgba(Vector3 v, float height)
		{
			int r = (int)(127.0f * v.X + 128.0f);
			int g = (int)(127.0f * v.Y + 128.0f);
			int b = (int)(127.0f * v.Z + 128.0f);
			int a = (int)(255.0f * height);
    
			return((a<<24) + (r<<16) + (g<<8) + (b<<0));
		}




		/// <summary>
		/// Initializes a vertex
		/// </summary>
		/// <returns>The initialized vertex</returns>
		private DotVertex InitializeVertex(float x, float y, float z, float tu, float tv)
		{
			DotVertex vertex = new DotVertex();
			Vector3 v = new Vector3(1,1,1);
			v.Normalize();

			vertex.p = new Vector3(x, y, z);
			vertex.diffuse = VectorToRgba(v, 1.0f);
			vertex.specular = 0x40400000;
			vertex.tu1 = tu; vertex.tv1 = tv;
			return vertex;
		}



		
        /// <summary>
        /// The window has been created, but the device has not been created yet.  
        /// Here you can perform application-related initialization and cleanup that 
        /// does not depend on a device.
        /// </summary>
        protected override void OneTimeSceneInitialization()
		{
			quadVerts[0] = InitializeVertex(-1.0f,-1.0f,-1.0f, 0.0f, 0.0f);
			quadVerts[1] = InitializeVertex(1.0f,-1.0f,-1.0f, 1.0f, 0.0f);
			quadVerts[2] = InitializeVertex(-1.0f, 1.0f,-1.0f, 0.0f, 1.0f);
			quadVerts[3] = InitializeVertex(1.0f, 1.0f,-1.0f, 1.0f, 1.0f);

			lightVector = new Vector3(0.0f, 0.0f, 1.0f);
		}



		/// <summary>
		/// Called once per frame, the call is the entry point for animating the scene.
		/// </summary>
		protected override void FrameMove()
		{
			if (System.Windows.Forms.Form.ActiveForm == this)
			{
				System.Drawing.Point pt = this.PointToClient(System.Windows.Forms.Cursor.Position);

				lightVector.X = -(((2.0f * pt.X) / device.PresentationParameters.BackBufferWidth) - 1);
				lightVector.Y = -(((2.0f * pt.Y) / device.PresentationParameters.BackBufferHeight) - 1);
				lightVector.Z = 0.0f;

				if (lightVector.Length() > 1.0f)
					lightVector.Normalize();
				else
					lightVector.Z = (float)Math.Sqrt(1.0f - lightVector.X*lightVector.X
						- lightVector.Y*lightVector.Y);
			}
		}




		/// <summary>
		/// Called once per frame, the call is the entry point for 3d rendering. This 
		/// function sets up render states, clears the viewport, and renders the scene.
		/// </summary>
		protected override void Render()
		{
			// Clear the viewport
			device.Clear(ClearFlags.Target , System.Drawing.Color.Black.ToArgb(), 1.0f, 0);

			device.BeginScene();

			// Store the light vector, so it can be referenced in TextureArgument.TFactor
			int factor = VectorToRgba(lightVector, 0.0f);
			device.RenderState.TextureFactor = factor;

			// Modulate the texture (the normal map) with the light vector (stored
			// above in the texture factor)
			device.TextureState[0].ColorArgument1 = TextureArgument.TextureColor;
			device.TextureState[0].ColorOperation = TextureOperation.DotProduct3;
			device.TextureState[0].ColorArgument2 = TextureArgument.TFactor;

			// If user wants to see the normal map, override the above renderstates and
			// simply show the texture
			if (isShowingNormalMap)
				device.TextureState[0].ColorOperation = TextureOperation.SelectArg1;

			// Select which normal map to use
			if (isUsingFileBasedTexture)
				device.SetTexture(0, fileBasedNormalMap);
			else
				device.SetTexture(0, customNormalMap);

			// Draw the bumpmapped quad
			device.VertexFormat = DotVertex.Format;
			device.DrawUserPrimitives(PrimitiveType.TriangleStrip, 2, quadVerts);

			// Output statistics
			drawingFont.DrawText(2,  1, System.Drawing.Color.Yellow, frameStats);
			drawingFont.DrawText(2, 20, System.Drawing.Color.Yellow, deviceStats);

			device.EndScene();
		}


		/// <summary>
		/// Create a file based normap map
		/// </summary>
		void CreateFileBasedNormalMap()
		{
			Texture fileBasedSource = null;
			try
			{
				// Load the texture from a file
				fileBasedSource = GraphicsUtility.CreateTexture(device, "earthbump.bmp", Format.A8R8G8B8);
				SurfaceDescription desc = fileBasedSource.GetLevelDescription(0);
				fileBasedNormalMap = new Texture(device, desc.Width, desc.Height, fileBasedSource.LevelCount, 0, Format.A8R8G8B8, Pool.Managed);
				TextureLoader.ComputeNormalMap(fileBasedNormalMap, fileBasedSource, 0, Channel.Red, 1.0f);
			}
			catch 
			{
				SampleException e = new MediaNotFoundException();
				HandleSampleException(e, ApplicationMessage.ApplicationMustExit);
				throw e;
			}
			finally
			{
				if (fileBasedSource != null)
					fileBasedSource.Dispose();
			}
		}




        /// <summary>
        /// Creates a custom normal map
        /// </summary>
        void CreateCustomNormalMap()
		{
			int localHeight = 512;
			int localWidth = 512;

			// Create a 32-bit texture for the custom normal map
			customNormalMap = new Texture(device, localWidth, localHeight, 1, 0, Format.A8R8G8B8, Pool.Managed);

			try
			{
				int[,] data = (int[,])customNormalMap.LockRectangle(typeof(int), 0, 0, localWidth, localHeight);
				// Fill each pixel
				for (int j=0; j<localHeight; j++)
				{
					for (int i=0; i<localWidth; i++)
					{
						float xp = ((5.0f*i) / (localWidth-1));
						float yp = ((5.0f*j) / (localHeight-1));
						float x  = 2*(xp-(float)Math.Floor(xp))-1;
						float y  = 2*(yp-(float)Math.Floor(yp))-1;
						float z  = (float)Math.Sqrt(1.0f - x*x - y*y);

						// Make image of raised circle. Outside of circle is gray
						if ((x*x + y*y) <= 1.0f)
						{
							Vector3 vector = new Vector3(x, y, z);
							data[j, i] = VectorToRgba(vector, 1.0f);
						}
						else
							data[j, i] = unchecked((int)0x80808080);
					}
				}
			}
			finally
			{
				// Make sure to unlock the rectangle even if there's an error
				customNormalMap.UnlockRectangle(0);
			}
		}



        
        /// <summary>
        /// The device has been created.  Resources that are not lost on
        /// Reset() can be created here -- resources in Pool.Managed,
        /// Pool.Scratch, or Pool.SystemMemory.  Image surfaces created via
        /// CreateImageSurface are never lost and can be created here.  Vertex
        /// shaders and pixel shaders can also be created here as they are not
        /// lost on Reset().
		/// </summary>
		protected override void InitializeDeviceObjects()
		{
			drawingFont.InitializeDeviceObjects(device);
			CreateFileBasedNormalMap();
			CreateCustomNormalMap();

			// Set the menu states
			mnuCustomTexture.Checked = !isUsingFileBasedTexture;
			mnuFileTexture.Checked = isUsingFileBasedTexture;
			mnuShowNormal.Checked = isShowingNormalMap;
		}




		/// <summary>
        /// The device exists, but may have just been Reset().  Resources in
        /// Pool.Default and any other device state that persists during
        /// rendering should be set here.  Render states, matrices, textures,
        /// etc., that don't change during rendering can be set once here to
        /// avoid redundant state setting during Render() or FrameMove().
		/// </summary>
		protected override void RestoreDeviceObjects(System.Object sender, System.EventArgs e)
		{

			// Set the transform matrices
			Vector3 vEyePt    = new Vector3(0.0f, 0.0f, 2.0f);
			Vector3 vLookatPt = new Vector3(0.0f, 0.0f, 0.0f);
			Vector3 vUpVec    = new Vector3(0.0f, 1.0f, 0.0f);
			Matrix matWorld, matView, matProj;

			matWorld = Matrix.Identity;
			matView = Matrix.LookAtLH(vEyePt, vLookatPt, vUpVec);
			float aspect = (float)device.PresentationParameters.BackBufferWidth / (float)device.PresentationParameters.BackBufferHeight;
			matProj = Matrix.PerspectiveFovLH((float)Math.PI/4, aspect, 1.0f, 500.0f);
			device.Transform.World = matWorld;
			device.Transform.View = matView;
			device.Transform.Projection = matProj;

			// Set any appropiate state
			device.RenderState.Lighting = false;
			device.SamplerState[0].MinFilter = TextureFilter.Linear;
			device.SamplerState[0].MagFilter = TextureFilter.Linear;
		}




		/// <summary>
		/// Called during device initialization, this code checks the device for some 
		/// minimum set of capabilities
		/// </summary>
		protected override bool ConfirmDevice(Caps caps, VertexProcessingType vertexProcessingType, 
			Format adapterFormat, Format backBufferFormat)
		{
			// All we need is the DotProduct texture operation
			return caps.TextureOperationCaps.SupportsDotProduct3;
		}




        /// <summary>
		/// Menu event handlers
		/// </summary>
		private void FileTextureClicked(object sender, EventArgs e)
		{
			isUsingFileBasedTexture = true;
			// Set the menu states
			mnuCustomTexture.Checked = !isUsingFileBasedTexture;
			mnuFileTexture.Checked = isUsingFileBasedTexture;
		}
		private void CustomTextureClicked(object sender, EventArgs e)
		{
			isUsingFileBasedTexture = false;
			// Set the menu states
			mnuCustomTexture.Checked = !isUsingFileBasedTexture;
			mnuFileTexture.Checked = isUsingFileBasedTexture;
		}
		private void ShowNormalClicked(object sender, EventArgs e)
		{
			isShowingNormalMap = !isShowingNormalMap;
			// Set the menu states
			mnuShowNormal.Checked = isShowingNormalMap;
		}

		

        
        /// <summary>
        /// The main entry point for the application.
        /// </summary>
        static void Main() 
        {
            using (MyGraphicsSample d3dApp = new MyGraphicsSample())
            {                                 
                if (d3dApp.CreateGraphicsSample())
                    d3dApp.Run();
            }
        }
	}
}
